### DISCUSSION TITLE
How to temporarily restrict focus to certain widgets?

### DISCUSSION THREAD

**Discussion by jspv** [Textualize/textual#1176] [Q&A] [answered]
***Use case***:  active menus, modal dialogs, etc.  - I want to ensure that only specified widgets can receive focus until I determine otherwise.  

The current approach I am pursuing is to identify all widgets that are focusable and selectively remove that focus (remembering the ones I've changed) and then rest when done.  It there a smarter way?

For reference, what I'm doing now is:

```python
     def push_focus(self):
        """remove focus for everything"""
        self._focus_save = self.focused
        self.set_focus(None)
        for widget in self.screen.focus_chain:
            self._focuslist.append(widget)
            widget.can_focus = False

    def pop_focus(self):
        """restore focus"""
        while len(self._focuslist) > 0:
            self._focuslist.pop().can_focus = True
        if self._focus_save is not None:
            self.set_focus(self._focus_save)
```

**Comment by davep** [comment, marked as answer]
As you'll have noticed, there's no builtin method of doing this and what you've got there looks like a good approach (I might personally be tempted to wrap it up in a class of its own, depending on how often it'd be used and in how many apps, but the core idea looks sound to me).

As Textual stands right now the more "Textuallly" method of doing it would be to use `Screen`s; so if you have some sort of modal action that you want on top of your main screen, etc, you'd implement that in a `Screen`, push that onto the screen stack, and then pop it back off when done.

As a simple illustration of this, see the help screen [in the Five by Five example](https://github.com/Textualize/textual/blob/main/examples/five_by_five.py) -- the main game grid there is done as buttons, and the help screen is modal; so doing that as a screen and popping it up over the main game screen means the user can't interact with the main game while the help is visible.

**Reply by jspv** [reply]
Thanks!  I did initially implement this using a screen, but didn't like the jarring full replacement of the screen, I preferred to have it be a more familiar popup dialog.  It will be wrapped up in a class of its own (widget actually).  It's working pretty well in my test code:  

![Screen Shot 2022-11-15 at 18 54 16](https://user-images.githubusercontent.com/6540751/202050350-84f11a0d-1bf8-4e65-8fb1-005b43d10812.jpg)

**Reply by davep** [reply]
Looks good! And, yeah, in that case I think you're taking the best approach here.
